File: fsharp-2017/src/Day05.fs
module Day05

open Util

let step (a:int[], i:int) = 
  let i' = i + a.[i]
  a.[i] <- a.[i] + 1
  i'

let doSteps (a:int[]) = 
  let mutable i = step (a, 0)
  let mutable steps = 1
  while i < Array.length a do
    i <- step (a, i)
    steps <- steps + 1
  steps

let run (instr:string) =
  let a = instr |> mapSplitLine int
  printfn "Day 5-1 answer: %A" <| doSteps a

File: js-pipe-fkit_2017/src/day05.js
const { strToInts } = require('./helpers')

const step = (idx, a) /*int,[int]*/ => {
  const i = idx + a[idx]
  a[idx] += 1
  return i
}

const step2 = (idx, a) /*int,[int]*/ => {
  const i = idx + a[idx]
  a[idx] += (a[idx] > 2)? -1 : 1
  return i
}

const doSteps = (stepper, ints) /*fn,[int]*/ => {
  const a = [...ints]
  let i = stepper(0, a)
  let steps = 1
  while (i < a.length) {
    i = stepper(i, a)
    steps += 1
  }
  return steps
} 

const run = (inStr, log) => {
  const ints = strToInts(inStr)
  log.p1( doSteps(step, ints))  // 372139
  log.p2( doSteps(step2, ints)) // 29629538
}
module.exports = {
  run,
  doSteps,
  step,
  step2,
}